File: apps/backend-python/app/services/user_service.py
"""
User service for user management operations.
"""
from typing import Optional
from datetime import datetime

from app.core.security import verify_password, get_password_hash, validate_password_strength
from app.models.user import User
from app.repositories.user_repository import UserRepository
from app.schemas.user import UserUpdate, ChangePasswordRequest, UserProfile


class UserService:
    """Service for user management operations."""
    
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository
    
    async def get_user_by_id(self, user_id: str) -> Optional[User]:
        """Get user by ID."""
        return await self.user_repository.find_by_id(user_id)
    
    async def get_user_by_email(self, email: str) -> Optional[User]:
        """Get user by email."""
        return await self.user_repository.find_by_email(email)
    
    async def update_user(self, user_id: str, update_data: UserUpdate) -> UserProfile:
        """Update user information."""
        user = await self.user_repository.find_by_id(user_id)
        if not user:
            raise ValueError("User not found")
        
        # Update only provided fields
        if update_data.name is not None:
            user.name = update_data.name
        
        user.updated_at = datetime.utcnow()
        
        await self.user_repository.save(user)
        
        return UserProfile(
            id=user.id,
            email=user.email,
            name=user.name,
            email_verified=user.email_verified,
            created_at=user.created_at,
        )
    
    async def change_password(
        self, user_id: str, change_password_data: ChangePasswordRequest
    ) -> None:
        """Change user password."""
        user = await self.user_repository.find_by_id(user_id)
        if not user:
            raise ValueError("User not found")
        
        # Verify current password
        if not verify_password(
            change_password_data.current_password, user.password_hash
        ):
            raise ValueError("Current password is incorrect")
        
        # Validate new password strength
        is_valid, errors = validate_password_strength(
            change_password_data.new_password
        )
        if not is_valid:
            raise ValueError(f"Password validation failed: {', '.join(errors)}")
        
        # Hash new password
        user.password_hash = get_password_hash(change_password_data.new_password)
        user.updated_at = datetime.utcnow()
        
        await self.user_repository.save(user)
    
    async def delete_user(self, user_id: str) -> None:
        """Delete user account."""
        user = await self.user_repository.find_by_id(user_id)
        if not user:
            raise ValueError("User not found")
        
        # Delete user
        deleted = await self.user_repository.delete(user_id)
        if not deleted:
            raise ValueError("Failed to delete user")
    
    async def get_user_profile(self, user_id: str) -> UserProfile:
        """Get user profile information."""
        user = await self.user_repository.find_by_id(user_id)
        if not user:
            raise ValueError("User not found")
        
        return UserProfile(
            id=user.id,
            email=user.email,
            name=user.name,
            email_verified=user.email_verified,
            created_at=user.created_at,
        )

File: apps/backend/src/services/user.service.ts
import { FastifyInstance } from 'fastify';
import { UserPostgresRepository } from '../repositories/postgres/user.repository';
import { PasswordService } from './password.service';
import { User } from '@prisma/client';

export interface UpdateUserDto {
  name?: string;
  profileImage?: string;
}

export interface ChangePasswordDto {
  currentPassword: string;
  newPassword: string;
}

export class UserService {
  private userRepository: UserPostgresRepository;
  private passwordService: PasswordService;

  constructor(private app: FastifyInstance) {
    this.userRepository = new UserPostgresRepository(app);
    this.passwordService = new PasswordService();
  }

  async findById(id: string): Promise<User | null> {
    return await this.userRepository.findById(id);
  }

  async updateProfile(id: string, dto: UpdateUserDto): Promise<User | null> {
    const user = await this.userRepository.findById(id);
    if (!user) {
      return null;
    }

    return await this.userRepository.update(id, dto);
  }

  async changePassword(id: string, dto: ChangePasswordDto): Promise<boolean> {
    const user = await this.userRepository.findById(id);
    if (!user) {
      throw new Error('사용자를 찾을 수 없습니다');
    }

    // 현재 비밀번호 확인
    const isCurrentPasswordValid = await this.passwordService.compare(
      dto.currentPassword,
      user.password
    );
    if (!isCurrentPasswordValid) {
      throw new Error('현재 비밀번호가 올바르지 않습니다');
    }

    // 새 비밀번호 유효성 검사
    const passwordValidation = this.passwordService.validate(dto.newPassword);
    if (!passwordValidation.isValid) {
      throw new Error(passwordValidation.errors.join(', '));
    }

    // 새 비밀번호 해시화 및 업데이트
    const hashedPassword = await this.passwordService.hash(dto.newPassword);
    const updated = await this.userRepository.update(id, {
      password: hashedPassword,
    });

    return updated !== null;
  }

  async deactivateUser(id: string): Promise<boolean> {
    const updated = await this.userRepository.update(id, {
      isActive: false,
    });
    return updated !== null;
  }
}